src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -653,9 +675,45 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
                         }
                     }
 
-                    compilation = UpdateCompilationWithNewReferencesAndRecordAssemblySymbols(compilation, newReferences, metadataReferenceToProjectId);
+                    if (!Enumerable.SequenceEqual(compilation.ExternalReferences, newReferences))
+                    {
+                        compilation = compilation.WithReferences(newReferences);
+                    }
 
-                    this.WriteState(new FinalState(State.CreateValueSource(compilation, solution.Services), compilation, hasSuccessfullyLoaded), solution);
+                    // Now we run generators; if we don't have a generator driver at all, we must try create one
+                    if (generatorDriver.GeneratorDriver == null)
+                    {
+                        var generators = this.ProjectState.AnalyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                        var additionalTexts = this.ProjectState.AdditionalDocumentStates.Values.SelectAsArray(a => (AdditionalText)new AdditionalTextWithState(a));
+                        var compilationFactory = this.ProjectState.LanguageServices.GetRequiredService<ICompilationFactoryService>();
+
+                        generatorDriver = new TrackedGeneratorDriver(
+                            compilationFactory.CreateGeneratorDriver(
+                                this.ProjectState.ParseOptions!,
+                                generators,
+                                additionalTexts));
+                    }
+
+                    // We will finalize the compilation by adding full contents here.
+                    // PROTOTYPE: allow finalize compilation to incrementally update a prior version
+                    var compilationWithoutGeneratedFiles = compilation;
+
+                    if (generatorDriver.GeneratorDriver != null)
+                    {
+                        // PROTOTYPE: make an API to expose these diagnostics
+                        generatorDriver.GeneratorDriver.RunFullGeneration(compilation, out compilation, out var diagnostics, cancellationToken);
```

Comment:
can this not just return a tuple :)

Reply:
how are these diagnostics reported?  i.e. this seems like a pri to get working.

Reply:
why are these diagnostics *outside* of hte compilation?  i would assume the new compilation instance would just have these internally and `.GetDiagnostics()` on it would return them...

Reply:
Filed https://github.com/dotnet/roslyn/issues/42818 to discuss further.

Reply:
Right now they're unreported indeed -- this PR was originally written prior to that being a thing, so we'll have to do that as a mop-up.